pkg/controller/externalippool/validate.go

```diff
@@ -83,47 +92,161 @@ func (c *ExternalIPPoolController) ValidateExternalIPPool(review *admv1.Admissio
-	} else {
-		if subnetInfo.PrefixLength <= 0 || subnetInfo.PrefixLength >= 128 {
-			return fmt.Sprintf("invalid prefixLength %d", subnetInfo.PrefixLength), false
+		var mask net.IPMask
+		if gatewayIP.To4() != nil {
+			if subnetInfo.PrefixLength <= 0 || subnetInfo.PrefixLength >= 32 {
+				return fmt.Sprintf("invalid prefixLength %d", subnetInfo.PrefixLength), false
+			}
+			mask = net.CIDRMask(int(subnetInfo.PrefixLength), 32)
+		} else {
+			if subnetInfo.PrefixLength <= 0 || subnetInfo.PrefixLength >= 128 {
+				return fmt.Sprintf("invalid prefixLength %d", subnetInfo.PrefixLength), false
+			}
+			mask = net.CIDRMask(int(subnetInfo.PrefixLength), 128)
+		}
+		subnet = &net.IPNet{
+			IP:   gatewayIP.Mask(mask),
+			Mask: mask,
 		}
-		mask = net.CIDRMask(int(subnetInfo.PrefixLength), 128)
 	}
-	subnet := &net.IPNet{
-		IP:   gatewayIP.Mask(mask),
-		Mask: mask,
+
+	// parse existing ip cidr and start-end ranges
+	existingCIDRs := make([]*net.IPNet, 0)
+	existingStartEndIPRanges := make([][2]net.IP, 0)
+	existingStartEndBigIntRanges := make([][2]*big.Int, 0)
+	for _, externalIPPool := range existingExternalIPPools {
+		for _, ipRange := range externalIPPool.Spec.IPRanges {
+			if ipRange.CIDR != "" {
+				cidr, _ := parseIPRangeCIDR(ipRange.CIDR)
+				existingCIDRs = append(existingCIDRs, cidr)
+			} else {
+				start, end, _ := parseIPRangeStartEnd(ipRange.Start, ipRange.End)
+				existingStartEndIPRanges = append(existingStartEndIPRanges, [2]net.IP{start, end})
+				existingStartEndBigIntRanges = append(existingStartEndBigIntRanges, [2]*big.Int{netutils.BigForIP(start), netutils.BigForIP(end)})
+			}
+		}
 	}
+
+	// validate ip cidr ranges
 	for _, ipRange := range ipRanges {
 		if ipRange.CIDR != "" {
-			_, cidr, err := net.ParseCIDR(ipRange.CIDR)
-			if err != nil {
-				return err.Error(), false
+			// parse cidr
+			cidr, errMsg := parseIPRangeCIDR(ipRange.CIDR)
+			if errMsg != "" {
+				return errMsg, false
 			}
-			if !ip.IPNetContains(subnet, cidr) {
-				return fmt.Sprintf("cidr %s must be a strict subset of the subnet", ipRange.CIDR), false
+
+			// validate if cidr is subset of given subnet info
+			if subnet != nil && !utilip.IPNetContains(subnet, cidr) {
+				return fmt.Sprintf("cidr %s must be a strict subset of the subnet %s", ipRange.CIDR, subnet.String()), false
 			}
-		} else {
-			start := net.ParseIP(ipRange.Start)
-			end := net.ParseIP(ipRange.End)
-			if !subnet.Contains(start) || !subnet.Contains(end) {
-				return fmt.Sprintf("IP range %s-%s must be a strict subset of the subnet", ipRange.Start, ipRange.End), false
+
+			// validate if cidr overlaps with any existing externalIPPool cidr
+			for _, existingCIDR := range existingCIDRs {
+				if cidr.Contains(existingCIDR.IP) || existingCIDR.Contains(cidr.IP) {
+					return fmt.Sprintf("cidr %s overlaps with cidr %s", cidr.String(), existingCIDR.String()), false
+				}
+			}
+
+			// validate if cidr overlaps with any existing externalIPPool start-end
```

Comment:
There are 4 such checks, is it simpler to just convert CIDR to `[2]*big.Int` to unify the validation, which should reduce the redundancy. To be able to return clear message including the original range string,  we could make existingStartEndBigIntRanges `map[string][2]*big.Int`.
```
// parse existing ip cidr and start-end ranges
existingStartEndBigIntRanges := make(map[string][2]*big.Int, 0)
for _, externalIPPool := range existingExternalIPPools {
	for _, ipRange := range externalIPPool.Spec.IPRanges {
		var start, end *big.Int
		if ipRange.CIDR != "" {
			start, end, _ = parseIPRangeCIDR(ipRange.CIDR)
			existingStartEndBigIntRanges[ipRange.CIDR] = [2]*big.Int{netutils.BigForIP(start), netutils.BigForIP(end)})
		} else {
			start, end, _ = parseIPRangeStartEnd(ipRange.Start, ipRange.End)
			existingStartEndBigIntRanges[fmt.Springf("%s-%s", ipRange.Start, ipRange.End)] = [2]*big.Int{netutils.BigForIP(start), netutils.BigForIP(end)})
		}
	}
}

for _, ipRange := range ipRanges {
	if ipRange.CIDR != "" {
		...
		for rangeStr, range := range existingStartEndBigIntRanges {
			if !(startBigInt.Cmp(range[1]) == 1 || endBigInt.Cmp(range[0]) == -1) {
			      ...
			}
		}
		existingStartEndBigIntRanges[ipRange.CIDR] = [2]*big.Int{startBigInt, end})
	} else {
	    ...
	}
}
```

Reply:
Sure, will unify the CIDR and start-end style ranges.